# Used Car Dataset Analysis


In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
The aim of this project is to clean the data and analyze the included used car listings.

# Objective

The objective of this project is to clean every column of this dataset so that it can be used for data analysis and machine learning purposes. This particular project would not be dealing with the analysis and machine learning prospects.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

# Renaming the columns

In [5]:
autos.rename(columns={'dateCrawled' : 'date_crawled', 'offerType' : 
                     'offer_type', 'vehicleType' : 'vehicle_type',
                     'yearOfRegistration' : 'registration_year',
                     'powerPS' : 'power_ps', 'monthOfRegistration' :
                     'registration_month', 'fuelType' : 'fuel_type',
                     'notRepairedDamage' : 'unrepaired_damage',
                     'dateCreated' : 'ad_created', 'nrOfPictures' :
                     'nr_of_pictures', 'postalCode' : 'postal_code',
                     'lastSeen' : 'last_seen'}, inplace=True)

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  object
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

# Brief info of every column

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

# Converting date_crawled from object to date data type

In [8]:
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'])

In [9]:
autos['date_crawled'].dtype

dtype('<M8[ns]')

# Exploring the column

In [10]:
autos["date_from_date_crawled"] = autos["date_crawled"].dt.date
value_counts = autos['date_from_date_crawled'].value_counts(normalize=True, dropna=False)
value_counts = value_counts.sort_index()
print(value_counts)

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_from_date_crawled, dtype: float64


It looks like the data was acquired through the whole month and the distribution of listings through each day seems to be fairly equal.

The next column is the name column which consists of the name of the car. This columns seems to have too many unique rows which indicates the seller named the car as the title post(describing the brand, model name, color etc.) at his own convenience. This column will not be very useful in the analysis part but we do not need to touch it.

Now let's see the seller and offer type columns

In [11]:
print(autos["seller"].value_counts())
print(autos["offer_type"].value_counts())

privat        49999
gewerblich        1
Name: seller, dtype: int64
Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64


Both these columns seems useless so let's drop these columns.

In [12]:
autos = autos.drop(["seller", "offer_type"], axis=1)

The next column is the price column which is usally a very important column

In [13]:
autos["price"].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,200      639
           ... 
$1,497        1
$79,980       1
$1,798        1
$10,949       1
$13,798       1
Name: price, Length: 2357, dtype: int64

Seems like "$" and "," are preventing it from becoming an integer data type.

In [14]:
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "")
autos["price"] = autos["price"].astype(int)
print(autos["price"].dtype)

int64


In [15]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

The minimum price seems to be zero and the maximum number seems to be hundred million which is a bit weird. As of now, no car in the world sells for that much. We will need to clear the price column to include only listings that are in a reasonable price range.

Let's see the cars which are the highest priced.

In [16]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

Now let's see the lowest priced cars.

In [17]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

It seems like there are many outliers in the data. We can remove all the rows in which the price of cars is listed between 30$ - $300000. 

In [18]:
autos = autos[autos["price"].between(30, 300000)]

In [19]:
autos["price"].describe()

count     48368.000000
mean       5898.537545
std        8800.554570
min          30.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      299000.000000
Name: price, dtype: float64

Seems like we didn't remove too many rows out of our dataset which is a good thing and the range seems reasonable too which further along will not skew the data analysis results.


In [27]:
autos["abtest"].value_counts(dropna=False)

test       24909
control    23459
Name: abtest, dtype: int64

The abtest column tells us that around 25k people were shown the new version of the platform while around the same people were shown the existing version of the platform.

In [26]:
autos["vehicle_type"].value_counts(dropna=False)

limousine     12569
kleinwagen    10557
kombi          8913
NaN            4491
bus            4024
cabrio         3007
coupe          2456
suv            1963
andere          388
Name: vehicle_type, dtype: int64

Most vehicle in our dataset on listings is a limousine

In [29]:
autos["registration_month"].value_counts(dropna=False)

3     4990
0     4383
6     4262
4     4028
5     4024
7     3847
10    3583
12    3364
9     3322
11    3310
1     3210
8     3117
2     2928
Name: registration_month, dtype: int64

The months in which the vehicle was registered seems pretty uniquely distributed.

In [30]:
autos["fuel_type"].value_counts(dropna=False)

benzin     29274
diesel     14337
NaN         3947
lpg          667
cng           72
hybrid        37
elektro       18
andere        16
Name: fuel_type, dtype: int64

The most common fuel type is benzin which is the german word for gasoline. We also have around 4k null rows where the fuel type isn't mentioned

In [31]:
autos["nr_of_pictures"].value_counts(dropna=False)

0    48368
Name: nr_of_pictures, dtype: int64

Seems like every row is just the number 0, which indicates no picture was uploaded to any posting which cannot be true. Looks like a failiure of scraping method used. We will delete this column since it has no useful value.

In [32]:
autos = autos.drop(["nr_of_pictures"], axis=1)

Now lets change the data type of odometer column and convert to int


In [33]:
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int)

In [34]:
autos.rename({"odometer" : "odometer_km"}, axis=1, inplace=True)

In [35]:
autos["odometer_km"].value_counts(dropna=False)

150000    31295
125000     5046
100000     2107
90000      1733
80000      1414
70000      1215
60000      1153
50000      1011
40000       815
5000        793
30000       778
20000       759
10000       249
Name: odometer_km, dtype: int64

Seems like the numbers are rounded which indicates the seller had to chose from a pre defined number.Additionally, there are more high mileage than low mileage vehicles.

In [40]:
print(autos["ad_created"].dtype)

object


In [41]:
autos['ad_created'] = pd.to_datetime(autos['ad_created'])

In [43]:
autos['ad_created'].dtype

dtype('<M8[ns]')

In [44]:
autos["date_from_ad_created"] = autos["ad_created"].dt.date
value_counts = autos['date_from_ad_created'].value_counts(normalize=True, dropna=False)
value_counts = value_counts.sort_index()
print(value_counts)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038848
2016-04-04    0.036884
2016-04-05    0.011785
2016-04-06    0.003246
2016-04-07    0.001240
Name: date_from_ad_created, Length: 76, dtype: float64


The data indicates there are some ads created almost nine months ago and can still be seen in the postings although most recent ads are the ones that gets the most traction.


In [45]:
autos['last_seen'] = pd.to_datetime(autos['last_seen'])
autos['last_seen'].dtype

dtype('<M8[ns]')

In [46]:
autos["date_from_last_seen"] = autos["last_seen"].dt.date
value_counts = autos['date_from_last_seen'].value_counts(normalize=True, dropna=False)
value_counts = value_counts.sort_index()
print(value_counts)

2016-03-05    0.001075
2016-03-06    0.004321
2016-03-07    0.005417
2016-03-08    0.007360
2016-03-09    0.009634
2016-03-10    0.010627
2016-03-11    0.012384
2016-03-12    0.023776
2016-03-13    0.008870
2016-03-14    0.012632
2016-03-15    0.015878
2016-03-16    0.016457
2016-03-17    0.028076
2016-03-18    0.007319
2016-03-19    0.015816
2016-03-20    0.020613
2016-03-21    0.020592
2016-03-22    0.021398
2016-03-23    0.018607
2016-03-24    0.019765
2016-03-25    0.019166
2016-03-26    0.016788
2016-03-27    0.015568
2016-03-28    0.020840
2016-03-29    0.022308
2016-03-30    0.024748
2016-03-31    0.023859
2016-04-01    0.022866
2016-04-02    0.024872
2016-04-03    0.025182
2016-04-04    0.024520
2016-04-05    0.125000
2016-04-06    0.221593
2016-04-07    0.132071
Name: date_from_last_seen, dtype: float64


The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [47]:
autos["registration_year"].describe()

count    48368.000000
mean      2004.776919
std         88.819202
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

In [49]:
(~autos["registration_year"].between(1900, 2016)).sum() / 48368

0.038785974197816736

Given that this is less than 4% of our data, we will remove these rows.

Let's remove those rows

In [50]:
autos = autos[autos["registration_year"].between(1900, 2016)]

# Conclusion

The dataset had some datatype problems which were solved. Some columns which were useless or gave false data were removed. And lastly, some rows which were definately giving wrong information were also removed. The dataset can now presumably be used to do analysis and find out some interesting stuff about the listings.
